### PR TITLE
Differential LR: Boost Specialized Heads (aft_srf, surface_refine, GSB)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The specialized correction heads (aft-foil SRF, surface refinement, Gap-Stagger Spatial Bias MLP) were added to a baseline with globally tuned LR=2e-4. These heads are zero-initialized — they start inert and must "wake up" during training. With the default LR, they receive the same update magnitude as the backbone, but they are starting from zero while the backbone starts from a pre-converged state (from cosine warmup).

**Theoretical motivation:** Standard transfer learning / fine-tuning practice applies higher LR to newly added layers vs frozen backbone. Here the backbone isn't frozen, but the specialized heads are at a structural disadvantage: they need to learn task-specific patterns (fore/aft surface corrections, geometry routing) on top of an already-tuned backbone. A 2x-3x LR boost gives them the gradient signal needed to converge faster within the 160-epoch budget.

**Specific mechanism:** The GSB spatial bias MLP currently lives in the `attn_params` group (LR = 1e-4 = base_lr * 0.5 for the attention group). This is HALF the effective LR vs the backbone — the GSB params are being under-trained. Raising GSB params to `base_lr * 2.0` = 4e-4 may unlock further gains from the mechanism that just delivered -3.0% p_tan.

### Step 1: Restructure optimizer parameter groups

Locate the optimizer setup in train.py (search for `attn_params`, `Lion([`). The current structure uses two groups: `attn_params` (lr * 0.5) and `other_params` (lr * 1.0).

**Change 1:** Extract `spatial_bias` params from `attn_params` into a new boosted group.
**Change 2:** Add surface_refine and aft_srf head params into the same boosted group.

```python
# Current (approximate):
attn_params  = [p for n, p in model.named_parameters() if any(k in n for k in ['Wqkv', 'temperature', 'slice_weight', 'attn_scale', 'spatial_bias'])]
other_params = [p for n, p in model.named_parameters() if p not in set(attn_params)]

# NEW:
_boost_keywords = ['spatial_bias']  # GSB MLP params get boosted LR
backbone_attn  = [p for n, p in model.named_parameters() if any(k in n for k in ['Wqkv', 'temperature', 'slice_weight', 'attn_scale']) and not any(k in n for k in _boost_keywords)]
backbone_other = [p for n, p in model.named_parameters() if not any(k in n for k in ['Wqkv', 'temperature', 'slice_weight', 'attn_scale', 'spatial_bias'])]
backbone_gsb   = [p for n, p in model.named_parameters() if 'spatial_bias' in n]

_head_lr_mult = cfg.head_lr_mult  # new config flag, default=2.0

base_opt = Lion([
    {'params': backbone_attn,  'lr': _base_lr * 0.5},
    {'params': backbone_other, 'lr': _base_lr},
    {'params': backbone_gsb,   'lr': _base_lr * _head_lr_mult},  # boosted
], weight_decay=cfg.weight_decay)
```

**Change 3:** When adding surface_refine and aft_srf head params to the optimizer (search for `add_param_group`), use the boosted LR:

```python
# For surface_refine head:
base_opt.add_param_group({'params': list(surface_refine_head.parameters()), 'lr': _base_lr * _head_lr_mult})

# For aft_srf head:
base_opt.add_param_group({'params': list(aft_srf_head.parameters()), 'lr': _base_lr * _head_lr_mult})
```

### Step 2: Add config flag

```python
# Config dataclass:
head_lr_mult: float = 1.0  # default 1.0 = no change

# argparse:
parser.add_argument('--head_lr_mult', type=float, default=1.0,
                    help='LR multiplier for specialized heads (aft_srf, surface_refine, GSB). Default=1.0 (no change).')
```

### Step 3: Run 3 configs × 1-2 seeds

Test two multiplier values to find the sweet spot:

```bash
# Config A: head_lr_mult=2.0, seed 42
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/head-lr2-s42" --wandb_group phase6/differential-lr \
  --head_lr_mult 2.0 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Config A: seed 73 (same, --seed 73, wandb_name "nezuko/head-lr2-s73")

# Config B: head_lr_mult=3.0, seed 42
# (same flags, --head_lr_mult 3.0, wandb_name "nezuko/head-lr3-s42")

# Config B: seed 73 (same, --seed 73, wandb_name "nezuko/head-lr3-s73")
```

If head_lr_mult=2.0 shows clear improvement, that's the winner; if it also helps at 3.0, report both. If 2.0 regresses, try 1.5 as a fallback.

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| mult=2.0 | 42 | | | | | |
| mult=2.0 | 73 | | | | | |
| **mult=2.0 avg** | — | | | | | |
| mult=3.0 | 42 | | | | | |
| mult=3.0 | 73 | | | | | |
| **mult=3.0 avg** | — | | | | | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

Also confirm the parameter groups are set up correctly by logging the LR for each group at epoch 1.

## Baseline

| Metric | Target |
|--------|--------|
| p_in   | < 13.05 |
| p_oodc | < 7.70  |
| **p_tan** | **< 28.60** |
| p_re   | < 6.55  |

W&B baseline: d7l91p0x (s42), j9btfx09 (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Differential LR for Specialized Heads

### Implementation
- Extracted `spatial_bias` (GSB) params from `attn_params` into a boosted group
- Applied `head_lr_mult` to GSB, surface_refine, and aft_srf head param groups
- Verified parameter groups at startup:
  - backbone_attn: 1.0e-04, backbone_other: 2.0e-04
  - GSB spatial_bias: 4.0e-04 (mult=2.0) / 6.0e-04 (mult=3.0)
  - refinement head: same as GSB
  - aft-foil SRF head: same as GSB

### Results (Surface MAE — Pressure)

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B ID |
|--------|------|------|--------|-------|------|--------|
| mult=2.0 | 42 | 13.507 | 7.898 | 28.770 | 6.495 | `jg8ummcf` |
| mult=2.0 | 73 | 13.469 | 7.945 | 30.194 | 6.577 | `mth1a9p8` |
| **mult=2.0 avg** | — | **13.488** | **7.922** | **29.482** | **6.536** | |
| mult=3.0 | 42 | 13.397 | 7.543 | 29.893 | 6.597 | `aadm4fz6` |
| mult=3.0 | 73 | 13.767 | 7.317 | 29.444 | 6.404 | `ltvxv838` |
| **mult=3.0 avg** | — | **13.582** | **7.430** | **29.669** | **6.501** | |
| **Baseline** | — | **13.05** | **7.70** | **28.60** | **6.55** | d7l91p0x, j9btfx09 |

### Analysis

**Neither config beats baseline on p_tan** (the key metric). Both regress on p_in and p_tan.

**Interesting finding: mult=3.0 significantly improves p_oodc** (7.43 vs 7.70 baseline, -3.5%). This is the OOD condition split (extreme AoA/gap/stagger) — boosting the GSB spatial bias MLP's LR helps it learn better geometry-conditioned routing for extreme conditions. However, this comes at the cost of p_in and p_tan regression.

**Why p_tan regressed:** The higher head LR may be causing the specialized heads to overfit to their specific correction patterns too quickly. With the cosine LR schedule, the heads see larger updates early on, and by the time the backbone has converged, the heads may have locked into suboptimal correction patterns. The baseline's uniform LR allows all components to co-adapt gradually.

**mult=2.0 vs mult=3.0:** Surprisingly, mult=3.0 is better on OOD metrics (p_oodc, p_re) while mult=2.0 is better on in-distribution (p_in) and p_tan. This suggests a trade-off: higher head LR → better OOD generalization via stronger specialized corrections, but worse in-distribution fit.

### Suggested Follow-ups
1. **Delayed head LR boost**: Use base_lr for the first N epochs, then switch to boosted LR — let the backbone converge first before accelerating heads
2. **Only boost GSB, not surface_refine/aft_srf**: The p_oodc improvement is likely from the GSB MLP; the other heads may not benefit from boosting
3. **Combine with a lower base attention LR**: The current attn_lr = 0.5 × base may be too aggressive alongside the boosted heads

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2163 because ADVISOR: Closing — differential LR regresses p_tan for both multiplier values.

**Results:**
- mult=2.0 avg: p_tan=29.48 (+3.1%), p_in=13.49 (+3.4%)
- mult=3.0 avg: p_tan=29.67 (+3.7%), p_in=13.58 (+4.1%), but p_oodc=7.43 (-3.5%)

**Interesting but insufficient:** The p_oodc improvement with mult=3.0 (-3.5%) shows the GSB MLP benefits from higher LR for OOD condition generalization. But the p_tan and p_in regressions are too large — the heads overfit their correction patterns too quickly with boosted LR, disrupting the co-adaptation with the backbone.

**Key insight:** The uniform LR enables gradual co-adaptation between backbone and specialized heads. Disrupting this balance hurts even if individual components benefit. The LR for specialized heads is not a bottleneck.

Good analysis, nezuko — the trade-off you identified (higher head LR → better OOD generalization but worse in-dist) is a useful finding. Moving you to a new experiment..
